### PR TITLE
Additional edits to Bitswap

### PR DESCRIPTION
**Pull request by edgo914**

### CODE REVIEW

**Comment by jessicaschilling** [comment]
@johnnymatthews - Can you please have a look? Thanks!

File: docs/concepts/bitswap.md [hunk 1]
```diff
@@ -11,4 +10,54 @@ related:
...
+The decision engine processes incoming messages that contain a wantlist. It checks the CID for each block in the wantlist against the local block-store and creates a task in the peer request queue for each block it finds. Task workers pull the tasks off the queue, retrieve the block requested, and send it off.
+
+The decision engine makes use of a ledger as a record of the aggregate data exchanged between two peers. An IPFS node maintains one ledger for each connected peer.
+
+### Additional references
+
+[Article: Swapping bits and distributing hashes on the decentralized web (Textile)](https://medium.com/textileio/swapping-bits-and-distributing-hashes-on-the-decentralized-web-5da98a3507)
+
+"About Bitswap" poster from the IPFS developer summit in Berlin in July 2018:
+!['About Bitswap' poster](https://user-images.githubusercontent.com/74178/43230914-f818dab2-901e-11e8-876b-73ba6a084f76.jpg "Bitswap-Poster_Berlin-July-2018")
```
Comment (johnnymatthews) [inline docs/concepts/bitswap.md]: This image needs to be updated and recreated in Photoshop/Illustrator at some point. That's outside the scope of this bounty though.

**Comment by johnnymatthews** [comment]
Great PR @edgo914, thanks for your help!

I've thrown in a few minor changes, but things look fairly solid from my end. I've asked one of the PL engineers to give the post a once over, just to make sure we're not missing anything major.

File: docs/concepts/bitswap.md (hunk 1 above)
Comment (jessicaschilling) [reply docs/concepts/bitswap.md]: @johnnymatthews You might want to file an issue in https://github.com/protocol/design-shop/ for that. Not sure where it'll fit into Agata's overall timeline, but probably worth getting it into her queue.

File: docs/concepts/bitswap.md [hunk 2]
```diff
@@ -11,4 +10,54 @@ related:
...
+### Message
+
+A single Bitswap message may contain any of the following content:
+
+-  **The sender’s wantlist.** This wantlist may either be the sender’s complete wantlist or just the changes to the sender’s wantlist that the receiver needs to know.
+-  **Data blocks.** These are blocks that have been requested (i.e., blocks in the sender’s wantlist, as far as the receiver is aware at the time of message receipt).
+
+### Session
+
+IPFS content blocks are often connected to each other through a [Merkle DAG](https://docs-beta.ipfs.io/concepts/merkle-dag/). If a sender's block requests are related, a mechanism called a _Bitswap session_ can optimize block requests sent to other peers. This mechanism can increase transfer speed and reduce the number of duplicate blocks on the network. Active peers are favored, as they are tracked relative to them having the requested blocks and how quickly they respond.
```
Comment (hsanjuan) [inline docs/concepts/bitswap.md]: The gist of a bitswap session is that if a remote peer(s) has one block, it is likely to have other blocks from the same DAG, so new blocks are preferentially requested from that/those peers (thus skipping more expensive paths to content discovery).

This is what this says, but language here is a bit convoluted and it took me a couple of reads.

File: docs/concepts/bitswap.md [hunk 3]
```diff
@@ -11,4 +10,54 @@ related:
...
+
+### Provider/Providing
+
+The provider system is responsible for announcing and reannouncing to the IPFS network that a node has content. If a node cannot find content with connected peers, it will query a content routing system (usually a form of [DHT](https://docs-beta.ipfs.io/concepts/dht/)) to locate a peer with the content. Bitswap routes these requests through the `ProviderQueryManager` system. As a node receives blocks, it announces them on the provided content routing system.
+
+### Deeper Examination
+
+A diagram showing the primary tasks of Bitswap and underlying components:
+
+![Bitswap task and component diagram](https://github.com/ipfs/go-bitswap/blob/master/docs/go-bitswap.png "Go-Bitswap")
```
Comment (hsanjuan) [inline docs/concepts/bitswap.md]: I am not sure if this is up to date or will stay up to date in the near future (cc. @dirkmc)

Comment (hsanjuan) [reply docs/concepts/bitswap.md]: Also, should note that this is specific to the go-bitswap implementation.

Comment (dirkmc) [reply docs/concepts/bitswap.md]: This is going to change with the next release of go-bitswap, and is not so relevant for other implementations of Bitswap. I'm not sure how useful it is for the audience that this document is aimed at to know about the architectural details of how Bitswap works. Probably the broader, conceptual picture outlined in the previous section is sufficient.

**Comment by edgo914** [comment]
Thanks for comments from @hsanjuan and @dirkmc. I suppose I wrote more than
needed here, but better to be able to take it away than not have enough.
Eh, I'm a coder, and I did notice the (vast) difference in functionality
when comparing the go vs js source. I will look through the references that
@dirkmc provided. Leaving now for a few days, so any additional comments I
will see next week, thanks.

**Comment by jessicaschilling** [comment]
Closing, as superseded by https://github.com/ipfs/ipfs-docs-v2/pull/92

**Pull request closed without merging**